kv/batch.go

```diff
@@ -0,0 +1,242 @@
+// The number of non-noop requests after truncation is returned along
+// with a closure which must be executed to undo the truncation, even
+// in case of an error.
+// TODO(tschottdorf): Consider returning a new BatchRequest, which has more
+// overhead in the common case of a batch which never needs truncation but is
+// less magical.
+func truncate(br *proto.BatchRequest, desc *proto.RangeDescriptor, from, to proto.Key) (func(), int, error) {
+	if !desc.ContainsKey(from) {
+		from = desc.StartKey
+	}
+	if !desc.ContainsKeyRange(desc.StartKey, to) || to == nil {
+		to = desc.EndKey
+	}
+	truncateOne := func(args proto.Request) (bool, []func(), error) {
+		if _, ok := args.(*proto.NoopRequest); ok {
+			return true, nil, nil
+		}
+		header := args.Header()
+		if !proto.IsRange(args) {
+			if len(header.EndKey) > 0 {
+				return false, nil, util.Errorf("%T is not a range command, but EndKey is set", args)
+			}
+			if !desc.ContainsKey(keys.KeyAddress(header.Key)) {
+				return true, nil, nil
+			}
+			return false, nil, nil
+		}
+		var undo []func()
+		key, endKey := header.Key, header.EndKey
+		keyAddr, endKeyAddr := keys.KeyAddress(key), keys.KeyAddress(endKey)
+		if keyAddr.Less(from) {
+			undo = append(undo, func() { header.Key = key })
+			header.Key = from
+			keyAddr = from
+		}
+		if !endKeyAddr.Less(to) {
+			undo = append(undo, func() { header.EndKey = endKey })
+			header.EndKey = to
+			endKeyAddr = to
+		}
+		// Check whether the truncation has left any keys in the range. If not,
+		// we need to cut it out of the request.
+		return !keyAddr.Less(endKeyAddr), undo, nil
+	}
+
+	var fns []func()
+	gUndo := func() {
+		for _, f := range fns {
+			f()
+		}
+	}
+
+	var numNoop int
+	for pos, arg := range br.Requests {
+		omit, undo, err := truncateOne(arg.GetValue().(proto.Request))
+		if omit {
+			numNoop++
+			nReq := &proto.RequestUnion{}
+			nReq.SetValue(&proto.NoopRequest{})
+			oReq := br.Requests[pos]
+			br.Requests[pos] = *nReq
+			posCpy := pos // for closure
+			undo = append(undo, func() {
+				br.Requests[posCpy] = oReq
+			})
+		}
+		fns = append(fns, undo...)
+		if err != nil {
+			return gUndo, 0, err
+		}
+	}
+	return gUndo, len(br.Requests) - numNoop, nil
+}
+
+// SenderFn is a function that implements a Sender.
+type senderFn func(context.Context, proto.BatchRequest) (*proto.BatchResponse, error)
+
+// SendBatch implements batch.Sender.
+func (f senderFn) SendBatch(ctx context.Context, ba proto.BatchRequest) (*proto.BatchResponse, error) {
+	return f(ctx, ba)
+}
+
+// A ChunkingSender sends batches, subdividing them appropriately.
+type chunkingSender struct {
+	f senderFn
+}
+
+// NewChunkingSender returns a new chunking sender which sends through the supplied
+// SenderFn.
+func newChunkingSender(f senderFn) client.BatchSender {
+	return &chunkingSender{f: f}
+}
+
+// SendBatch implements Sender.
+// TODO(tschottdorf): We actually don't want to chop EndTransaction off for
```

Comment:
What is this comment referring to? There's nothing related to EndTransaction here now.